src/mc-rangeopts.c

```diff
@@ -324,6 +351,162 @@ bool mc_RangeOpts_appendMax(const mc_RangeOpts_t *ro,
+        }
+        *bitsOut = 64 - (uint32_t)mc_count_leading_zeros_u64(out.max);
+        return true;
+    } else if (valueType == BSON_TYPE_DATE_TIME) {
+        int64_t value;
+        mc_optional_int64_t rmin, rmax;
+        if (ro->min.set) {
+            value = bson_iter_date_time(&ro->min.value);
+            rmin = OPT_I64(value);
+            rmax = OPT_I64(bson_iter_date_time(&ro->max.value));
+        } else {
+            value = 0;
+            rmin.set = false;
+            rmax.set = false;
+        }
+        mc_getTypeInfo64_args_t args = {value, rmin, rmax};
+        mc_OSTType_Int64 out;
+        if (!mc_getTypeInfo64(args, &out, status)) {
+            return false;
+        }
+        *bitsOut = 64 - (uint32_t)mc_count_leading_zeros_u64(out.max);
+        return true;
+    } else if (valueType == BSON_TYPE_DOUBLE) {
+        double value;
+        mc_optional_double_t rmin, rmax;
+        mc_optional_uint32_t prec = ro->precision;
+        if (ro->min.set) {
+            value = bson_iter_double(&ro->min.value);
+            rmin = OPT_DOUBLE(value);
+            rmax = OPT_DOUBLE(bson_iter_double(&ro->max.value));
+
+        } else {
+            value = 0;
+            rmin.set = false;
+            rmax.set = false;
+        }
+        mc_getTypeInfoDouble_args_t args = {value, rmin, rmax, prec};
+        mc_OSTType_Double out;
+        if (!mc_getTypeInfoDouble(args, &out, status)) {
+            return false;
+        }
+        *bitsOut = 64 - (uint32_t)mc_count_leading_zeros_u64(out.max);
+        return true;
+    }
+#if MONGOCRYPT_HAVE_DECIMAL128_SUPPORT
+    else if (valueType == BSON_TYPE_DECIMAL128) {
+        mc_dec128 value;
+        mc_optional_dec128_t rmin, rmax;
+        mc_optional_uint32_t prec = ro->precision;
+        if (ro->min.set) {
+            value = mc_dec128_from_bson_iter(&ro->min.value);
+            rmin = OPT_MC_DEC128(value);
+            rmax = OPT_MC_DEC128(mc_dec128_from_bson_iter(&ro->max.value));
+        } else {
+            value = MC_DEC128_ZERO;
+            rmin.set = false;
+            rmax.set = false;
+        }
+        mc_getTypeInfoDecimal128_args_t args = {value, rmin, rmax, prec};
+        mc_OSTType_Decimal128 out;
+        if (!mc_getTypeInfoDecimal128(args, &out, status)) {
+            return false;
+        }
+        *bitsOut = 128 - (uint32_t)mc_count_leading_zeros_u128(out.max);
+        return true;
+    }
+#endif
+    CLIENT_ERR("unsupported BSON type: %s for range", mc_bson_type_to_string(valueType));
+    return false;
+}
+
+bool mc_RangeOpts_appendTrimFactor(const mc_RangeOpts_t *ro,
+                                   bson_type_t valueType,
+                                   const char *fieldName,
+                                   bson_t *out,
+                                   mongocrypt_status_t *status) {
+    BSON_ASSERT_PARAM(ro);
+    BSON_ASSERT_PARAM(fieldName);
+    BSON_ASSERT_PARAM(out);
+    BSON_ASSERT(status || true);
+
```

Comment:
What would be an acceptable way to do something like this with libmongocrypt? In the server, we have a feature flag which bounds the use of trimFactor, but I wasn't sure about what the equivalent would be in libmongocrypt.

Reply:
I do not think there is an established pattern aside from `mongocrypt_setopt_fle2v2`. I expect it can be a private and short-lived flag. See https://github.com/mongodb/libmongocrypt/commit/68a41291bf9839dc1423ce77c0b97e6ca54c244f https://github.com/mongodb/libmongocrypt/commit/49109dca07e259ef2cfa8f7e5fc976f82280ee61 on [this branch](https://github.com/mongodb/libmongocrypt/compare/master...kevinAlbs:libmongocrypt:marksg07/server-85740?expand=1) for one possible way.

Reply:
Thanks so much for this! I've merged those commits.